go/core/background_action.go

```diff
@@ -0,0 +1,188 @@
+// Copyright 2025 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package core
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/firebase/genkit/go/internal/registry"
+	"github.com/invopop/jsonschema"
+)
+
+// Operation represents a background task operation
+type Operation struct {
+	Action   string         `json:"action,omitempty"`   // The action that created this operation
+	ID       string         `json:"id"`                 // Unique identifier for tracking
+	Done     bool           `json:"done,omitempty"`     // Whether the operation is complete
+	Output   any            `json:"output,omitempty"`   // Result when done
+	Error    string         `json:"error,omitempty"`    // Error if failed
+	Metadata map[string]any `json:"metadata,omitempty"` // Additional info
+}
+
+// BackgroundModelOptions holds configuration for defining a background model
+type BackgroundModelOptions struct {
+	Versions       []string           // Known version names for this model
+	Supports       any                // Capabilities this model supports
+	ConfigSchema   *jsonschema.Schema // Custom options schema for this model
+	Label          string             // Descriptive name for this model
+	SupportsCancel bool               // Whether the model supports cancellation
+}
+
+// BackgroundStartFunc starts a background operation
+type BackgroundStartFunc[In, Out any] func(ctx context.Context, input In) (*Operation, error)
+
+// BackgroundCheckFunc checks the status of a background operation
+type BackgroundCheckFunc[Out any] func(ctx context.Context, operation *Operation) (*Operation, error)
+
+// BackgroundCancelFunc cancels a background operation
+type BackgroundCancelFunc[Out any] func(ctx context.Context, operation *Operation) (*Operation, error)
+
+// BackgroundAction interface represents a background action
+type BackgroundAction[In, Out any] interface {
+	Start(ctx context.Context, input In) (*Operation, error)
+	Check(ctx context.Context, operation *Operation) (*Operation, error)
+	Cancel(ctx context.Context, operation *Operation) (*Operation, error)
+	SupportsCancel() bool
+	Name() string
+}
+
+// backgroundActionImpl implements BackgroundAction
+type backgroundActionImpl[In, Out any] struct {
+	startAction    *ActionDef[In, *Operation, struct{}]
+	checkAction    *ActionDef[*Operation, *Operation, struct{}]
+	cancelAction   *ActionDef[*Operation, *Operation, struct{}]
+	name           string
+	supportsCancel bool
```

Comment:
Do we even need this? Can't we just check for the presence of `cancelAction` like we're already doing anyway?